refactor: Make custom map visitor publicly accessible

nmote opened the pull request:
This moves the nonstandard parts of of the mapper into AST_generic. Now, the mapper in `Map_AST`'s only responsibility is to override methods and call the appropriate client-supplied function. This makes it possible for callsites to start using the custom visitor directly, at which point we can delete `Map_AST.mk_visitor`.

Test plan: Automated tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

nmote commented on libs/ast_generic/AST_generic.ml:
```diff
@@ -2405,3 +2395,48 @@ let is_metavar_name s = Common.( =~ ) s "^\\(\\$[A-Z_][A-Z_0-9]*\\)$"
+    method! visit_expr env x =
+      let ekind = self#visit_expr_kind env x.e in
+      (* TODO? reuse the e_id or create a new one? *)
+      e ekind
```
This calls `e`, so I had to put it below the definition. I moved `iter_no_id_info` too, for consistency.

emjin commented on libs/ast_generic/AST_generic.ml:
```diff
@@ -2405,3 +2395,48 @@ let is_metavar_name s = Common.( =~ ) s "^\\(\\$[A-Z_][A-Z_0-9]*\\)$"
+ * new callsite, you should consider whether the nonstandard behavior here makes
+ * sense for your use case. Think about using the autogenerated AST_generic.map
+ * directly. *)
+class virtual ['self] map_custom =
```
I'd call this `Map_AST_default` instead or something similar

nmote replied:
Hmm, maybe I'll call it `map_legacy`? I'm not really sure exactly what to call it. Its defining characteristic is that it exists to mimic the previous behavior.

emjin commented on libs/ast_generic/AST_generic.ml:
```diff
@@ -2405,3 +2395,48 @@ let is_metavar_name s = Common.( =~ ) s "^\\(\\$[A-Z_][A-Z_0-9]*\\)$"
+    method! visit_id_info _env _info = ()
+  end
+
+(* This is based on the old behavior of Map_AST.mk_visitor. If you are adding a
```
To someone new "the old behavior of Map_AST.mk_visitor" wouldn't mean much. I would find it more helpful to say "This is tailored to Map_AST, see comment in Map_AST about why it makes certain choices" (if such a comment exists).

In contrast the comment above is great because it explains what's special about iter_no_id_info

nmote replied:
I will elaborate a bit, but I'm going to delete most, if not all of `Map_AST` shortly so referring to a comment there won't help much.

emjin reviewed (approved):
lgtm

nmote committed:
refactor: Make custom map visitor publicly accessible

This moves the nonstandard parts of of the mapper into AST_generic. Now,
the mapper in `Map_AST`'s only responsibility is to override methods and
call the appropriate client-supplied function. This makes it possible
for callsites to start using the custom visitor directly, at which point
we can delete `Map_AST.mk_visitor`.

Test plan: Automated tests
